### PR TITLE
Improve shops performance take 2

### PR DESCRIPTION
**Pull request by sauloperez**
#### What? Why?

Moves #3099 forward.

This is most of the stuff I had on my stash since last time I worked on #3099. It's just part of the refactoring needed to remove another one of the various N+1s that take place in `GET /shops`.

This is aligned with what I did in #3151, where an N+1 is removed by moving the data loading to the controller, or #3257 where moving the data loading brings some clarity.

Unlike #3151, here loading product and producer properties in a single query turns to be tougher. I had to stop at `#active` and give it some more thought. There's more on the backstage, but this is what it's shippable for now.

There won't be much performance improvement if any.

#### What should we test?

The distributed properties shown in `/shops` should keep working as expected. That is, those that belong to producers or products that belong or belonged to an order cycle.

#### Release notes

Refactor CachedEnterpriseSerializer's `#distributed_properties` to make it easier to remove an N+1 on the properties table in the future.

Changelog Category: Changed

### CODE REVIEW

File: app/serializers/api/property_serializer.rb [hunk 1]
```diff
@@ -1,4 +1,7 @@
 class Api::PropertySerializer < ActiveModel::Serializer
+  # TODO: stick to :presentation or :name not both because the are duplicate
```
Comment (luisramos0) [inline app/serializers/api/property_serializer.rb]: imo, code should never ever have TODOs. TODOs belong in backlogs. So I think this should be either 1. a comment about how the code works ":presentation or :name are duplicates and could be one property only" or 2. an issue in GH

File: app/serializers/api/cached_enterprise_serializer.rb [hunk 2]
```diff
@@ -81,25 +81,42 @@ def supplied_properties
...
-
+        Spree::Property.currently_sold_by(enterprise)
       else
-        product_properties = Spree::Property.ever_sold_by(object)
-        producer_property_ids = ProducerProperty.ever_sold_by(object).pluck(:property_id)
+        Spree::Property
+          .joins(products: { variants: { exchanges: :order_cycle } })
+          .merge(Exchange.outgoing)
+          .merge(Exchange.to_enterprise(enterprise))
+          .select('DISTINCT spree_properties.*')
```
Comment (luisramos0) [inline app/serializers/api/cached_enterprise_serializer.rb]: I dont get this part. you just moved the code from Spree::Property.ever_sold_by to here, right?

Comment (sauloperez) [reply app/serializers/api/cached_enterprise_serializer.rb]: Yep, that's it. Essentially, `#distributed_properties` issues few queries to the properties table but this makes things more complex than they should be. It's simpler to just ask the DB about properties that we care for than having to deal with `ProducerProperty` on one hand, `product_properties` on the other, then mapping their presentation attribute, etc. We tend to forget that SQL is declarative and end up implementing map-reduce.

A first step towards relying on SQL is to inline the queries to find how they can be merged together. Does it make sense?

File: app/serializers/api/property_serializer.rb (hunk 1 above)
Comment (sauloperez) [reply app/serializers/api/property_serializer.rb]: Ups, sorry my mistake. I left it there to turn it into something else. I share the opinion about TODOs

Comment (sauloperez) [reply app/serializers/api/property_serializer.rb]: I created https://github.com/openfoodfoundation/openfoodnetwork/issues/3313

File: app/serializers/api/cached_enterprise_serializer.rb (hunk 2 above)
Comment (luisramos0) [reply app/serializers/api/cached_enterprise_serializer.rb]: yes it makes sense. but I wouldnt move/duplicate the SQL here if I was not changing it.

File: spec/controllers/shops_controller_spec.rb [hunk 3]
```diff
@@ -15,4 +15,25 @@
...
     expect(response.body).to have_content(invisible_distributor.name)
   end
+
+  it 'renders distributed properties' do
+    duplicate_property = create(:property, presentation: 'dairy')
+    producer = create(:supplier_enterprise, properties: [duplicate_property])
+    property = create(:property, presentation: 'dairy')
+    product = create(:product, properties: [property])
+    producer.supplied_products << product
+
```
Comment (kristinalim) [inline spec/controllers/shops_controller_spec.rb]: If you want to doing a smoke test here, maybe it is more appropriate to test that producer and product properties are covered, than checking that there are no duplicates.

**Comment by luisramos0** [comment]
@sauloperez moving to "in dev", pls move to code review when you think is ready ;-)

File: spec/controllers/shops_controller_spec.rb (hunk 3 above)
Comment (sauloperez) [reply spec/controllers/shops_controller_spec.rb]: The absence of duplicates is already covered in the serializer specs (and it does work) but I added individual tests for product and producer properties in the controller as well.

**Comment by sauloperez** [comment]
I believe I fixed all comments and it's ready to be tested. Agree @luisramos0 and @kristinalim ?

File: spec/serializers/api/cached_enterprise_serializer_spec.rb [hunk 4]
```diff
@@ -27,28 +27,26 @@
...
-      order_cycle.exchanges << incoming_exchange
-      order_cycle.exchanges << outgoing_exchange
-
-      order_cycle.save
+      create(
+        :simple_order_cycle,
+        coordinator: shop,
+        suppliers: [producer],
+        distributors: [shop],
+        variants: [product.variants]
```
[the remark this answers is not in the document]
Comment (kristinalim) [reply spec/serializers/api/cached_enterprise_serializer_spec.rb]: @sauloperez This seems to still work anyway, but shouldn't this be `variants: product.variants`? Minor change needed.

**Comment by kristinalim** [comment]
@sauloperez From the PR description, this doesn't close the issue yet. I disconnected this from the issue on Zenhub, and then moved to Test Ready. :-)

**Comment by RachL** [comment]
Properties are looking good. Moving to ready to go https://docs.google.com/document/d/1fBnxEQiuA3FOXMmUJbX04F35NhMnL4XQjFLo21gJEWo/edit#heading=h.hj26umg3kyxb
